drivers/audio/intel_dmic.c

```diff
@@ -0,0 +1,1456 @@
+
+/* Macros to set bit(s) */
+#define SET_BIT(b, x)		(((x) & 1) << (b))
+#define SET_BITS(b_hi, b_lo, x)	\
+	(((x) & ((1 << ((b_hi) - (b_lo) + 1)) - 1)) << (b_lo))
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+/* queue size to hold buffers in process */
+#define DMIC_BUF_Q_LEN		2
+
+#define DMIC_REG_RD(reg)	(*((volatile u32_t *)(PDM_BASE + (reg))))
+#define DMIC_REG_WR(reg, val)	\
+	(*((volatile u32_t *)(PDM_BASE + (reg))) = (val))
+#define DMIC_REG_UPD(reg, mask, val)		\
+	DMIC_REG_WR((reg), (DMIC_REG_RD((reg)) & ~(mask)) | ((val) & (mask)))
+
+#define DCACHE_INVALIDATE(addr, size)		\
+	xthal_dcache_region_invalidate(addr, size)
+
+struct _stream_data {
+	struct k_msgq in_queue;
+	struct k_msgq out_queue;
+	void *in_msgs[DMIC_BUF_Q_LEN];
+	void *out_msgs[DMIC_BUF_Q_LEN];
+	struct k_mem_slab *mem_slab;
+	size_t block_size;
+};
+
+/* DMIC private data */
+static struct _dmic_pdata {
+	enum dmic_state state;
+	u16_t fifo_a;
+	u16_t fifo_b;
+	u16_t mic_en_mask;
+	u8_t num_streams;
+	u8_t reserved;
+	struct _stream_data streams[DMIC_MAX_STREAMS];
+	struct device *dma_dev;
+} dmic_private;
+
+static inline void dmic_parse_channel_map(u32_t channel_map_lo,
+		u32_t channel_map_hi, u8_t channel, u8_t *pdm, enum pdm_lr *lr);
+static inline u8_t dmic_parse_clk_skew_map(u32_t skew_map, u8_t pdm);
+static void dmic_stop(void);
+
+/* This function searches from vec[] (of length vec_length) integer values
+ * of n. The indexes to equal values is returned in idx[]. The function
+ * returns the number of found matches. The max_results should be set to
+ * 0 (or negative) or vec_length get all the matches. The max_result can be set
+ * to 1 to receive only the first match in ascending order. It avoids need
+ * for an array for idx.
+ */
+int find_equal_int16(s16_t idx[], s16_t vec[], int n, int vec_length,
+	int max_results)
+{
+	int nresults = 0;
+	int i;
+
+	for (i = 0; i < vec_length; i++) {
+		if (vec[i] == n) {
+			idx[nresults++] = i;
+			if (nresults == max_results)
+				break;
+		}
+	}
+
+	return nresults;
+}
+
+/* Return the smallest value found in the vector */
+s16_t find_min_int16(s16_t vec[], int vec_length)
+{
+	int i;
+	int min = vec[0];
+
+	for (i = 1; i < vec_length; i++)
+		min = (vec[i] < min) ? vec[i] : min;
+
+	return min;
+}
+
+/* Return the largest absolute value found in the vector. Note that
+ * smallest negative value need to be saturated to preset as s32_t.
+ */
+s32_t find_max_abs_int32(s32_t vec[], int vec_length)
+{
+	int i;
+	s64_t amax = (vec[0] > 0) ? vec[0] : -vec[0];
+
+	for (i = 1; i < vec_length; i++) {
+		amax = (vec[i] > amax) ? vec[i] : amax;
+		amax = (-vec[i] > amax) ? -vec[i] : amax;
+	}
+
+	return SATP_INT32(amax); /* Amax is always a positive value */
+}
+
+/* Count the left shift amount to normalize a 32 bit signed integer value
+ * without causing overflow. Input value 0 will result to 31.
+ */
+int norm_int32(s32_t val)
```

Comment:
Some architectures have dedicated instructions for that. Please try to use __builtin_clz() to speedup this operation on these architectures.

Reply:
OK will look into this.